// @discordx/koa/build/cjs/decorators/classes/DRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DReqeuest = void 0;
const Method_js_1 = require("./Method.js");
class DReqeuest extends Method_js_1.Method {
    _api;
    _name;
    _description;
    _type;
    _path;
    _middleWares = [];
    get api() {
        return this._api;
    }
    set api(value) {
        this._api = value;
    }
    get type() {
        return this._type;
    }
    get name() {
        return this._name;
    }
    get description() {
        return this._description;
    }
    get path() {
        return this._path;
    }
    get middlewares() {
        return this._middleWares;
    }
    set middlewares(value) {
        this._middleWares = value;
    }
    constructor(data) {
        super();
        this._name = data.name;
        this._description = data.description;
        this._type = data.type;
        this._path = data.path;
    }
    setName(value) {
        this._name = value;
    }
    setDescription(value) {
        this._description = value;
    }
    static create(data) {
        return new DReqeuest(data);
    }
}
exports.DReqeuest = DReqeuest;
