// @discordx/koa/build/cjs/decorators/classes/DRouter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DRouter = void 0;
const tslib_1 = require("tslib");
const di_1 = require("@discordx/di");
const internal_1 = require("@discordx/internal");
const router_1 = (0, tslib_1.__importDefault)(require("@koa/router"));
class DRouter extends internal_1.Decorator {
    _api;
    _name;
    _description;
    _router;
    get api() {
        return this._api;
    }
    set api(value) {
        this._api = value;
    }
    get name() {
        return this._name;
    }
    get description() {
        return this._description;
    }
    get instance() {
        return di_1.DIService.instance.getService(this.from);
    }
    get router() {
        return this._router;
    }
    constructor(data) {
        super();
        this._name = data.name;
        this._description = data.description;
        this._router = new router_1.default(data.options);
    }
    setName(value) {
        this._name = value;
    }
    setDescription(value) {
        this._description = value;
    }
    static create(data) {
        return new DRouter(data);
    }
}
exports.DRouter = DRouter;
