// @discordx/koa/build/cjs/decorators/decorators/Request.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Options = exports.Unlink = exports.Link = exports.Head = exports.Delete = exports.All = exports.Post = exports.Get = void 0;
const DRequest_js_1 = require("../classes/DRequest.js");
const metadata_js_1 = require("../../logic/metadata.js");
function RequestMethod(method, path) {
    return function (target, key) {
        metadata_js_1.MetadataStorage.instance.addRequest(DRequest_js_1.DReqeuest.create({
            name: key,
            path: path ?? `/${key}`,
            type: method,
        }).decorate(target.constructor, key, target[key]));
    };
}
function Get(path) {
    return RequestMethod("GET", path);
}
exports.Get = Get;
function Post(path) {
    return RequestMethod("POST", path);
}
exports.Post = Post;
function All(path) {
    return RequestMethod("ALL", path);
}
exports.All = All;
function Delete(path) {
    return RequestMethod("DELETE", path);
}
exports.Delete = Delete;
function Head(path) {
    return RequestMethod("HEAD", path);
}
exports.Head = Head;
function Link(path) {
    return RequestMethod("LINK", path);
}
exports.Link = Link;
function Unlink(path) {
    return RequestMethod("UNLINK", path);
}
exports.Unlink = Unlink;
function Options(path) {
    return RequestMethod("OPTIONS", path);
}
exports.Options = Options;
